src/WorkspaceAction.ts

```
@@ -0,0 +1,151 @@
+  }
+
+  filterWorkspacesImpendingAutostop(workspace: Workspace) {
+    // a workspace is eligible for autostop if the last build was successful,
+    // and the workspace is started,
+    // and it has a deadline
+    if (
+      workspace.latest_build.job.status !== "succeeded" ||
+      workspace.latest_build.transition !== "start" ||
+      !workspace.latest_build.deadline
+    ) {
+      return false
+    }
+
+    const hourMilli = 1000 * 60 * 60
+    // return workspaces with a deadline that is in 1 hr or less
+    return Math.abs(new Date().getTime() - new Date(workspace.latest_build.deadline).getTime()) <= hourMilli
+  }
+
+  filterWorkspacesImpendingDeletion(workspace: Workspace) {
+    if (!workspace.deleting_at) {
+      return
+    }
+
+    const dayMilli = 1000 * 60 * 60 * 24
+
+    // return workspaces with a deleting_at  that is 24 hrs or less
+    return Math.abs(new Date().getTime() - new Date(workspace.deleting_at).getTime()) <= dayMilli
+  }
+
+  transformWorkspaceObjects(workspace: Workspace, deadlineField?: string) {
+    // the below line is to satisfy TS; we should always pass a deadlineField, e.g
+    // workspace,deleting_at or workspace.latest_build.deadline
+    if (!deadlineField) {
+      return { workspace, wasNotified: true, impendingActionDeadline: "" }
+    }
+    const wasNotified =
+      this.#workspacesApproachingAutostop.find((wn) => wn.workspace.id === workspace.id)?.wasNotified ?? false
+    const impendingActionDeadline = formatDistanceToNowStrict(new Date(deadlineField))
+    return { workspace, wasNotified, impendingActionDeadline }
+  }
+
+  async pollGetWorkspaces() {
+    let errorCount = 0
+    this.#fetchWorkspacesInterval = setInterval(async () => {
+      try {
+        const workspacesResult = await getWorkspaces({ q: "owner:me" })
+        this.#ownedWorkspaces = workspacesResult.workspaces
+        this.seedNotificationLists()
+        this.notifyAll()
+      } catch (error) {
```

Thoughts on whether we should raise these errors in some way?  Even if just logging it somewhere out of sight.  I think the plugin has an output channel somewhere.